libflux/src/semantic/nodes.rs

```diff
@@ -0,0 +1,567 @@
+// Module nodes contains the nodes for the semantic graph.
+//
+// NOTE(affo): At this stage the nodes are a clone of the AST nodes with some type information added.
+//  Nevertheless, new node types allow us to decouple this step of compilation from the parsing.
+//  This is of paramount importance if we decide to add responsibilities to the semantic analysis and
+//  change it independently from the parsing bits.
+//  Uncommented node types are a direct port of the AST ones.
+//
+// NOTE(affo): With respect to the Go implementation, the Extern* nodes have been eliminated,
+//  because they where used only for including type inference information in the semantic nodes.
+//  We now implement a different strategy here, where type information is embedded in the Expressions.
+//
+// NOTE(affo): With respect to the Go implementation, the FunctionExpression node has beeen simplified
+//  to provide piped arguments and defaults via methods.
+
+extern crate chrono;
+use crate::ast;
+use crate::semantic::types;
+
+use chrono::prelude::DateTime;
+use chrono::Duration;
+use chrono::FixedOffset;
+use std::vec::Vec;
+
+pub type Location<'a> = &'a ast::SourceLocation;
+
+// TypeInformation represents the type information for a semantic node.
+#[derive(Debug, PartialEq, Clone)]
+pub struct TypeInformation {
+    pub mono_type: types::MonoType,
+}
```

Comment:
Is there a reason for wrapping the monotype vs attaching the monotype to the node directly?

Reply:
It was mostly to implement `update` once and forall.

Do you preder to remove the method and attache only the monotype?